[CI] Speed up CI test and fix flaky test

yzang2019 opened the pull request:
<!--
### Contribution Checklist

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.

**(The sections below can be removed for hotfixes of typos)**
-->


### Motivation

KoP test is too slow, and there are couple of test very flaky and always timeout.

### Modifications

Parallelize KoP uni tests with multiple jobs after regroup the tests into corresponding packages
Add a testing matrix for github workflow to spin up one job for each group of tests
Increase timeout to fix some flaky tests

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/streamnative/kop?pullRequest=1925) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/streamnative/kop?pullRequest=1925) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/streamnative/kop?pullRequest=1925) it.</sub>

yzang2019 commented on tests/src/test/java/io/streamnative/pulsar/handlers/kop/TransactionTest.java:
```diff
@@ -93,25 +93,25 @@ protected static Object[][] produceConfigProvider() {
     }
 
-    @Test(timeOut = 1000 * 10, dataProvider = "produceConfigProvider")
+    @Test(timeOut = 1000 * 30, dataProvider = "produceConfigProvider")
```
I've run these locally and it takes 22 seconds

yzang2019 commented on tests/src/test/java/io/streamnative/pulsar/handlers/kop/MetricsProviderTest.java:
```diff
@@ -314,7 +314,7 @@ public void testUpdateGroupId() {
     }
 
-    @Test(timeOut = 20000, expectedExceptions = KeeperException.NoNodeException.class)
+    @Test(timeOut = 60000, expectedExceptions = KeeperException.NoNodeException.class)
```
Tested this locally which takes usually 20-30 seconds to finish

BewareMyPower commented:
I don't think it speeds up the CI test actually.

<img width="844" alt="image" src="https://github.com/streamnative/kop/assets/18204803/f60006c3-7813-479e-bc4c-97b9f3c44245">

You can see the `module-test`, i.e. the tests under the `tests/` directory, took the most of the time of CI (44 minutes), while the other tests only took 14 minutes in summary.

See another CI example in https://github.com/streamnative/kop/pull/1914:

<img width="867" alt="image" src="https://github.com/streamnative/kop/assets/18204803/68fbda48-0003-403e-9122-b6139ebccd94">

How could you say this PR speeds up the CI test?

gaoran10 commented:
> I don't think it speeds up the CI test actually.

Maybe we need to add some test groups for tests and use different ci jobs to test various test groups.

BewareMyPower commented:
> Maybe we need to add some test groups for tests 

+1

yzang2019 commented:
> I don't think it speeds up the CI test actually.
> 
> <img alt="image" width="844" src="https://user-images.githubusercontent.com/18204803/249029078-f60006c3-7813-479e-bc4c-97b9f3c44245.png">
> You can see the `module-test`, i.e. the tests under the `tests/` directory, took the most of the time of CI (44 minutes), while the other tests only took 14 minutes in summary.
> 
> See another CI example in #1914:
> 
> <img alt="image" width="867" src="https://user-images.githubusercontent.com/18204803/249029325-68fbda48-0003-403e-9122-b6139ebccd94.png">
> How could you say this PR speeds up the CI test?

Yeah, you are right, I only realized this after I split it into multiple jobs, I think next step is to split them into groups

yzang2019 commented:
@BewareMyPower @Demogorgon314 I've addressed comments, total test time dropped from 50 minutes to around 10 minutes now, could you take another look?

gaoran10 commented on tests/src/test/java/io/streamnative/pulsar/handlers/kop/e2e/BasicEndToEndKafkaTest.java:
```diff
@@ -60,27 +62,25 @@ public BasicEndToEndKafkaTest() {
 
-        @Cleanup
-        final KafkaProducer<String, String> kafkaProducer = newKafkaProducer();
+        @Cleanup final KafkaProducer<String, String> kafkaProducer = newKafkaProducer();
```
Could we retain the code format?

gaoran10 reviewed (commented):
Great work! LGTM, left a trivial comment.

BewareMyPower reviewed (approved):
Great work!
